File: recipes/subms-timer-wheel/rust/examples/growth_main.rs
//! Storage-growth capture for the timer wheel: under continuous schedule+tick
//! churn, the pending-timer index (`id_to_slot`) must reach a bounded steady
//! state and stay there. If `tick` fired timers but forgot to evict their ids,
//! the index would climb every round - a slow leak a percentile can't see. This
//! measures the index size round over round and gates it flat.
//!
//! Emits the stable subms growth JSON on stdout.
//!
//! ```sh
//! cat <<EOF | cargo run --release --example growth_main --features harness
//! rounds=50
//! num_slots=256
//! ops_per_round=20000
//! EOF
//! ```

use std::collections::BTreeMap;
use std::io::{self, Read};
use std::process::ExitCode;

use subms::{SubMsGrowthClass, SubMsGrowthRecipe, grow, growth_to_json};
use subms_timer_wheel::TimerWheel;

// Rough per-pending-entry heap cost: the id->slot map entry plus the slot's
// Entry<u64> (id + rounds + value + flag).
const ENTRY_BYTES: u64 = 48;

struct WheelChurn {
    wheel: TimerWheel<u64>,
    num_slots: usize,
    rounds: usize,
    ops_per_round: usize,
    seq: u64,
}

impl SubMsGrowthRecipe for WheelChurn {
    fn name(&self) -> &str {
        "subms-timer-wheel"
    }
    fn op_name(&self) -> &str {
        "schedule"
    }
    fn rounds(&self) -> usize {
        self.rounds
    }
    fn ops_per_round(&self) -> usize {
        self.ops_per_round
    }
    fn op(&mut self, _round: usize, i: usize) {
        // Schedule one timer somewhere in the next rotation, then advance the hand
        // one tick (firing anything now due). Over many ops the in-flight set
        // reaches a steady size; a leaking tick would let it grow without bound.
        let delay = (i % (self.num_slots - 1)) + 1;
        self.wheel.schedule(delay, self.seq);
        self.seq += 1;
        let _ = self.wheel.tick();
    }
    fn memory_bytes(&mut self) -> u64 {
        self.wheel.pending() as u64 * ENTRY_BYTES
    }
    fn live_bytes(&mut self) -> u64 {
        // The genuinely-in-flight timers are the live set; a correct wheel holds
        // exactly those, so resident == live.
        self.wheel.pending() as u64 * ENTRY_BYTES
    }
    fn structures(&mut self) -> Vec<(String, u64)> {
        vec![("pending".to_string(), self.wheel.pending() as u64)]
    }
    fn expected(&self) -> (SubMsGrowthClass, f64) {
        // The pending index must plateau at its steady size, not climb round over
        // round - climbing would mean fired ids are never evicted.
        (SubMsGrowthClass::PlateauBounded, 1.5)
    }
}

fn parse_usize(map: &BTreeMap<String, String>, key: &str, default: usize) -> usize {
    map.get(key)
        .and_then(|v| v.trim().parse().ok())
        .unwrap_or(default)
}

fn main() -> ExitCode {
    let mut raw = String::new();
    if io::stdin().read_to_string(&mut raw).is_err() {
        eprintln!("growth_main: failed to read stdin");
        return ExitCode::FAILURE;
    }
    let mut map = BTreeMap::new();
    for line in raw.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some((k, v)) = line.split_once('=') {
            map.insert(k.trim().to_string(), v.trim().to_string());
        }
    }
    let rounds = parse_usize(&map, "rounds", 50);
    let num_slots = parse_usize(&map, "num_slots", 256).max(4);
    let ops_per_round = parse_usize(&map, "ops_per_round", 20_000);

    let mut recipe = WheelChurn {
        wheel: TimerWheel::new(num_slots),
        num_slots,
        rounds,
        ops_per_round,
        seq: 0,
    };
    let report = grow(&mut recipe, "rust");

    if growth_to_json(&report, &mut io::stdout().lock()).is_err() {
        eprintln!("growth_main: failed to write json");
        return ExitCode::FAILURE;
    }
    ExitCode::SUCCESS
}

File: recipes/subms-timer-wheel/java/src/main/java/com/submillisecond/recipes/timer/GrowthMain.java
package com.submillisecond.recipes.timer;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

import com.submillisecond.perf.SubMsGrowth;
import com.submillisecond.perf.SubMsGrowthRecipe;

/**
 * Storage-growth capture for the timer wheel: under continuous schedule+tick
 * churn, the pending-timer index ({@code idToSlot}) must reach a bounded steady
 * state and stay there. If {@code tick} fired timers but forgot to evict their
 * ids, the index would climb every round - a slow leak a percentile can't see.
 * This measures the index size round over round and gates it flat.
 *
 * <p>Emits the stable subms growth JSON on stdout. Mirror of the Rust
 * {@code examples/growth_main.rs}, reading the same stdin key=value config, so
 * the two curves are directly comparable.
 *
 * <pre>
 * printf 'rounds=50\nnum_slots=256\nops_per_round=20000\n' \
 *   | java -cp target/classes:... com.submillisecond.recipes.timer.GrowthMain
 * </pre>
 */
public final class GrowthMain {

    /**
     * Rough per-pending-entry heap cost: the HashMap node (32) with its boxed
     * Long id (24) and boxed Integer slot (16), plus the slot's Entry (32), its
     * slot-list array element (4), and the boxed Long payload (24).
     *
     * <p>132 against the Rust port's 48 for the same pending timer: Rust stores
     * the id and slot inline in a hash table and the payload inline in the slot,
     * where the JVM boxes all three and threads them through a linked node. Both
     * ports still hold exactly the live set, so the plateau verdict agrees even
     * though the byte counts do not.
     */
    private static final long ENTRY_BYTES = 32 + 24 + 16 + 32 + 4 + 24;

    private GrowthMain() {}

    private static final class WheelChurn implements SubMsGrowthRecipe {
        private final TimerWheel<Long> wheel;
        private final int numSlots;
        private final int rounds;
        private final int opsPerRound;
        private long seq;

        WheelChurn(int numSlots, int rounds, int opsPerRound) {
            this.wheel = new TimerWheel<>(numSlots);
            this.numSlots = numSlots;
            this.rounds = rounds;
            this.opsPerRound = opsPerRound;
        }

        @Override public String name() {
            return "subms-timer-wheel";
        }

        @Override public String opName() {
            return "schedule";
        }

        @Override public int rounds() {
            return rounds;
        }

        @Override public int opsPerRound() {
            return opsPerRound;
        }

        @Override public void op(int round, int i) {
            // Schedule one timer somewhere in the next rotation, then advance the
            // hand one tick (firing anything now due). Over many ops the in-flight
            // set reaches a steady size; a leaking tick would let it grow without
            // bound.
            long delay = (i % (numSlots - 1)) + 1;
            wheel.schedule(delay, seq);
            seq++;
            wheel.tick();
        }

        @Override public long memoryBytes() {
            return (long) wheel.pending() * ENTRY_BYTES;
        }

        @Override public long liveBytes() {
            // The genuinely-in-flight timers are the live set; a correct wheel holds
            // exactly those, so resident == live.
            return (long) wheel.pending() * ENTRY_BYTES;
        }

        @Override public Map<String, Long> structures() {
            return Map.of("pending", (long) wheel.pending());
        }

        @Override public SubMsGrowth.GrowthClass expectedClass() {
            return SubMsGrowth.GrowthClass.PLATEAU_BOUNDED;
        }

        @Override public double expectedBound() {
            // The pending index must plateau at its steady size, not climb round over
            // round - climbing would mean fired ids are never evicted.
            return 1.5;
        }
    }

    private static int parseInt(Map<String, String> cfg, String key, int fallback) {
        String v = cfg.get(key);
        if (v == null) {
            return fallback;
        }
        try {
            return Integer.parseInt(v.trim());
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    public static void main(String[] args) throws Exception {
        Map<String, String> cfg = new HashMap<>();
        try (BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8))) {
            String line;
            while ((line = in.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }
                int eq = line.indexOf('=');
                if (eq > 0) {
                    cfg.put(line.substring(0, eq).trim(), line.substring(eq + 1).trim());
                }
            }
        }

        var recipe = new WheelChurn(
                Math.max(4, parseInt(cfg, "num_slots", 256)),
                parseInt(cfg, "rounds", 50),
                parseInt(cfg, "ops_per_round", 20_000));

        Writer out = new OutputStreamWriter(System.out, StandardCharsets.UTF_8);
        SubMsGrowth.growthToJson(SubMsGrowth.grow(recipe, "java"), out);
        out.flush();
    }
}
